### PR TITLE
renew license api (issue #1347)

### PR DESCRIPTION
**Pull request by odedhutzler**
Hi,
The purpose of this pull request is to provide renewal license api for stored content.
It gets an offline uri (e.g "`offline:manifest/9`") and returns the offline manifest object (with the new expiration timestamp).

The flow is pretty much as I described [here](https://github.com/google/shaka-player/issues/1347)

My two main concerns are:
1. I exposed two properties in drm_engine (maybe there is a better way to do it - active_sessions + mediaKeys).
2. is there a simpler way to create the new manifest? what i am running the whole '`loadInternals`' flow, cause i guess i need the initData etc.

would be happy to hear what you think.

Thanks,
Oded

### CODE REVIEW

**Comment by ismena** [comment]
Thanks for the PR, @odedhutzler!

@joeyparrish Joey, I'm assigning this to you since you were the one to discuss the solution.

**Comment by odedhutzler** [comment]
hey @joeyparrish, any thoughts on this one? :)

**Comment by ismena** [comment]
@odedhutzler sorry for the delay, I will raise the visibility of this with @joeyparrish 

Also, let me ask our expert on offline to take a look. @TheModMaker, pretty please! :)

Please note that we're currently refactoring and redesigning parts of our offline implementation which might affect your PR.

**Comment by TheModMaker** [comment]
I'm concerned that you are making a request to `originalManifestUri`, which may have changed or be expired now.  You are assuming that the manifest is unchanged, which may not always be true.  Plus, the app may need to set up authentication to make the manifest request but they may not be expecting it.

Maybe a better plan would be to store the init (pssh) data in the offline manifest.  Then we can recreate the sessions using the original init data without needing to make a manifest request or having to load anything.  Unfortunately this will require changing the database a little and it would have to wait until #1248 is done.

**Comment by joeyparrish** [comment]
I agree with @TheModMaker.  For this, I think you should store the init data in the database instead of refetching the manifest.  See `externs/shaka/offline.js` for `shakaExtern.ManifestDB`.

Also, as @TheModMaker points out, your work may need to be adjusted once #1248 is complete, as we are refactoring the internals of `Storage` quite a bit.

**Comment by joeyparrish** [comment]
@vaage: JFYI, as this may intersect with your work on #1248.

**Comment by odedhutzler** [comment]
@joeyparrish @TheModMaker thanks for the review.

@vaage would be happy to be part of it when you get to it.

Thanks!

**Comment by TheModMaker** [comment]
#1248 is closed and we have finished the refactorings to Storage.  If you are still interested in this, please rebase and fix any style errors with our new style rules.

As for design, you could actually just keep the init data in [Storage](https://github.com/google/shaka-player/blob/master/lib/offline/storage.js#L726) and update the [ManifestConverter](https://github.com/google/shaka-player/blob/master/lib/offline/manifest_converter.js#L64) to not keep the init data if we already have offline sessions.  Then you just need to use `ManifestDb.drmInfos` to get the init data and use DrmEngine to delete the old sessions and request the new licences.  Finally you'll need to update the database with the new session IDs.

File: lib/offline/storage.js [hunk 1]
```diff
@@ -100,6 +100,90 @@ shaka.offline.Storage = function(player) {
...
+    this.filterAllPeriods_(data.manifest.periods);
+    this.pendingContent_ = OfflineUtils.createStoredContentFromManifest(
+        oldManifestDB_.originalManifestUri,
+        data.manifest,
+        0,  // start with a size of 0
+        oldManifestDB_.appMetadata);
+    newManifestDB_ = this.createOfflineManifest_(
+        oldManifestDB_.originalManifestUri, oldManifestDB_.appMetadata);
+    newManifestDB_.periods = storedPeriods;
+    this.drmEngine_.getActiveSessions()[0].session.close();
```
Comment (mariomc) [inline lib/offline/storage.js]: Careful. The close is a promise and you need to wait for it to resolve first.

**Comment by joeyparrish** [comment]
Closing due to inactivity.  If you need to reopen this issue, just put `@shaka-bot reopen` in a comment.  Thanks!

**Pull request closed without merging**